core/src/main/java/org/apache/calcite/rel/rules/HyperGraph.java

```diff
@@ -401,69 +453,67 @@ public RexNode extractJoinCond(RelNode left, RelNode right, List<HyperEdge> edge
   }
 
   /**
-   * Before starting enumeration, add Project on every input, make all field name unique.
-   * Convert the HyperEdge condition from RexInputRef to RexInputFieldName
+   * Restore the projection order of the final result to the original plan.
+   *
+   * @param resultOrder the node order of the final result
+   * @param rowTypeList rowType of the final result
+   * @return  list of RexInputRef
    */
-  public void convertHyperEdgeCond(RelBuilder builder) {
-    int fieldIndex = 0;
-    List<RelDataTypeField> fieldList = rowType.getFieldList();
-    for (int nodeIndex = 0; nodeIndex < inputs.size(); nodeIndex++) {
-      RelNode input = inputs.get(nodeIndex);
-      List<RexNode> projects = new ArrayList<>();
-      List<String> names = new ArrayList<>();
-      for (int i = 0; i < input.getRowType().getFieldCount(); i++) {
-        projects.add(
-            new RexInputRef(
-            i,
-            fieldList.get(fieldIndex).getType()));
-        names.add(fieldList.get(fieldIndex).getName());
-        fieldIndex++;
+  public List<RexNode> restoreProjectionOrder(
+      ImmutableList<Integer> resultOrder,
+      List<RelDataTypeField> rowTypeList) {
+    Map<Integer, Integer> relativePositionInNode = new HashMap<>();
+    int fieldCount = 0;
+    for (int resultIndex : resultOrder) {
+      relativePositionInNode.put(resultIndex, fieldCount);
+      if (!LongBitmap.isOverlap(notProjectInputs, LongBitmap.newBitmap(resultIndex))) {
+        fieldCount += inputs.get(resultIndex).getRowType().getFieldCount();
       }
-
-      builder.push(input)
-          .project(projects, names, true);
-      replaceInput(nodeIndex, builder.build());
     }
-
-    RexShuttle inputRef2inputNameShuttle = new RexShuttle() {
-      @Override public RexNode visitInputRef(RexInputRef inputRef) {
-        int index = inputRef.getIndex();
-        return new RexInputFieldName(
-            fieldList.get(index).getName(),
-            fieldList.get(index).getType());
+    List<RexNode> projects = new ArrayList<>();
+    for (int inputIndex = 0; inputIndex < inputs.size(); inputIndex++) {
+      if (LongBitmap.isOverlap(notProjectInputs, LongBitmap.newBitmap(inputIndex))) {
+        continue;
       }
-    };
 
-    for (int i = 0; i < edges.size(); i++) {
-      HyperEdge edge = edges.get(i);
-      RexNode convertCond = edge.getCondition().accept(inputRef2inputNameShuttle);
-      HyperEdge convertEdge =
-          new HyperEdge(
-              edge.getLeftNodeBitmap(),
-              edge.getRightNodeBitmap(),
-              edge.getJoinType(),
-              convertCond);
-      edges.set(i, convertEdge);
+      for (int i = 0; i < inputs.get(inputIndex).getRowType().getFieldCount(); i++) {
+        Integer fieldOffset = relativePositionInNode.get(inputIndex);
+        if (fieldOffset == null) {
+          throw new DpHyp.DphypOrHyperGraphException(
+              "The result order loses the " + inputIndex + "-th input");
```

Comment:
looses?
is this an assertion failure?
Then you should probably use the AssertionFailure exception.